### PR TITLE
**Fix:** Datepicker component: open the datepicker component on the correct month and year

### LINKED ISSUE #909: Open the DatePicker on the correct page if `max` is set
## Summary

Actually the `DatePicker` component always open on the current date, even if we have the `max` props set.

It's very annoying so let's fix this! 😄

### PR DESCRIPTION
**Pull request by rimildeyjsr**
Add the getStartMonthYearInWidget(props) method to Datepicker.tsx to set start month, year the datepicker should be opened to, based on the availability of start, end, max, min props. If no props are available, the current date is used.
Fixes #909

# Summary

Fixes #909

The datepicker component always opens on the current month and year when the start props is not available. Ideally, it should open on the month and year of any one of the props available, and should fallback to the current month and year only when none of the props are available.

# Related issue
#909 

# To be tested

Me
- [ ] No error or warning in the console on `localhost:6060`

Tester 1

- [x] Things look good on the demo.
  

Tester 2

- [ ] Things look good on the demo.

### CODE REVIEW

File: src/DatePicker/DatePicker.tsx [hunk 1]
```diff
@@ -78,6 +71,36 @@ class DatePicker extends React.Component<DatePickerProps, State> {
     }
   }
 
+  public getStartMonthYearInWidget(props?: DatePickerProps) {
+    // set start month and year of the widget based on the availability of start, end, max, or min props.
+    // if none of the props are available, set it to the current month and year.
+    const validatedProps = props || this.props
```
Comment (TejasQ) [inline src/DatePicker/DatePicker.tsx]: So this function is going to be recomputed on every render. This isn't ideal. 

I think to make it a little more performant, you could move this function to [DatePicker.utils.ts](https://github.com/contiamo/operational-ui/blob/1b5389bae8c376761e43c9c14fca802892b634c9/src/DatePicker/DatePicker.utils.ts) and call it from inside here, passing props to it.

Does that make sense?

Comment (rimildeyjsr) [reply src/DatePicker/DatePicker.tsx]: Yup that does! will make the changes and add them in!

**Comment by rimildeyjsr** [comment]
@TejasQ - Moved the `getStartMonthYearInWidget()` function out of the class and added unit tests for the `getStartMonthYearInWidget()`

Let me know if there are any issues 😃

**Comment by TejasQ** [comment]
Also @rimildeyjsr did you test this locally?
image

**Comment by rimildeyjsr** [comment]
@TejasQ - I had thought about subsets, but I included all so as to not miss any edge case. Any minor case missed would lead it back to having the current date in the date picker. Also, subsets quickly got complicated so I used the if-else ladder.

Can we merge?  😅

**Comment by rimildeyjsr** [comment]
@TejasQ - there are errors in the console, but the same errors are also there for the `master` branch. No new warnings or errors were introduced after my changes

**Comment by TejasQ** [comment]
Really good work, @rimildeyjsr! Amazing!

**Comment by rimildeyjsr** [comment]
@TejasQ - thanks! I learned a lot! Also, first time writing unit tests 😄

**Comment by TejasQ** [comment]
@rimildeyjsr weird – this worked on the CI but broke on `master`. If you checkout `master` and `yarn jest`, you'll see the unit test is broken. Would you like to fix it or shall I?

**Comment by rimildeyjsr** [comment]
@TejasQ - I will check it out!

**Comment by rimildeyjsr** [comment]
@TejasQ - I found the issue, the months and days are zero-indexed, hence the tests were failing. For eg, the unit test expected June(6) to return 5 (because 0 based) but was expecting 6. How should I add the code, do I raise a new PR?

**Comment by TejasQ** [comment]
Yup.

**Comment by TejasQ** [comment]
I still don’t know how the tests passed on CI for the previous PR.

**Comment by rimildeyjsr** [comment]
@TejasQ - Fix for the broken unit tests

https://github.com/contiamo/operational-ui/pull/1130
